src/v/raft/recovery_stm.cc

```diff
@@ -382,25 +416,86 @@ ss::future<> recovery_stm::handle_install_snapshot_reply(
     }
 
     // snapshot received by the follower, continue with recovery
-    (*meta)->match_index = _ptr->_last_snapshot_index;
-    (*meta)->next_index = model::next_offset(_ptr->_last_snapshot_index);
+    (*meta)->match_index = _last_included_snapshot_index;
+    (*meta)->next_index = model::next_offset(_last_included_snapshot_index);
     return close_snapshot_reader();
 }
 
-ss::future<> recovery_stm::install_snapshot() {
+ss::future<> recovery_stm::install_snapshot(required_snapshot_type s_type) {
+    const auto learner_start_offset = _ptr->get_learner_start_offset();
     // open reader if not yet available
-    auto f = _snapshot_reader != nullptr ? ss::now() : open_snapshot_reader();
-
-    return f.then([this]() mutable {
-        // we are outside of raft operation lock if snapshot isn't yet ready we
-        // have to wait for it till next recovery loop
-        if (!_snapshot_reader) {
-            _stop_requested = true;
-            return ss::now();
+    if (!_snapshot_reader) {
+        if (
+          s_type == required_snapshot_type::on_demand
+          && learner_start_offset > _ptr->start_offset()) {
+            co_await take_on_demand_snapshot(
+              model::prev_offset(*learner_start_offset));
+        } else {
+            co_await open_current_snapshot();
         }
+    }
 
-        return send_install_snapshot_request();
-    });
+    // we are outside of raft operation lock if snapshot isn't yet ready we
+    // have to wait for it till next recovery loop
+    if (!_snapshot_reader) {
+        _stop_requested = true;
+        co_return;
+    }
+    co_return co_await send_install_snapshot_request();
+}
+
+ss::future<>
+recovery_stm::take_on_demand_snapshot(model::offset last_included_offset) {
+    vlog(
+      _ctxlog.debug,
+      "creating on demand snapshot with last included offset: {}",
+      last_included_offset);
+
+    _last_included_snapshot_index = last_included_offset;
```

Comment:
updating this here seems a bit confusing.. isn't it better to update after send_install_snapshot_request() succeeds or some such?

Reply:
i will rename the variable to better reflect its purpose, i was confused myself when looking into it after some time. In general it is the offset of snapshot currently being delivered to the follower.